### PR TITLE
fixing multiple WPIMULT records

### PR DESCRIPTION
**Pull request by GitPaean**
An attempt to fix the running with multiple WIPMULT records.

The key modification here is that if `multiple records` within the `same report step` to a well has all defaulted and connection and completion information, only the `last` record will apply.

### CODE REVIEW

**Comment by GitPaean** [comment]
jenkins build this please

File: src/opm/input/eclipse/Schedule/Well/Well.cpp [hunk 1]
```diff
@@ -64,6 +64,9 @@ namespace {
         if (item.get<int>(0) == 0)
             return true;
 
+        if (item.get<int>(0) < 0)
```
Comment (GitPaean) [inline src/opm/input/eclipse/Schedule/Well/Well.cpp:67]: This is something needs to be discussed. 

For WELOPEN and WPIMULT, the default I, J, K values are negative
For COMLUMP, the default I, J, K values are 0. 

The observation is that for WPIMULT, the following record will apply to all connections
```
WPIMULT
  'WEL' 0.9 0 0 0 /
/
```
but the behavoir shows the commercial simulator **does not** think it is defaulted. 

But the following record, 
```
WPIMULT
  'WEL' 0.9 -1 -1 -1 /
/
```
the commercial simulator treats it as **defaulted** and also apply to all the connections.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
A test is added. So the PR is ready for review.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
The user that requested this task would like us to handle the situation with multiple `WPIMULT` keywords within the same report step.  So the PR is back to draft again. Crossing different keywords is kind of challenging the current way of handling schedule keywords.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
Now with the PR, the situations that multiple WIPMULT keywords within the same report step are handled. The results look good and test were added. But I think there might be suggestions about how we should actually handle it, especially for the interface of function `handleKeyword`.

@bska if you can give some quick review, it will be much appreciated.

File: opm/input/eclipse/Schedule/Schedule.hpp [hunk 2]
```diff
@@ -515,7 +516,8 @@ namespace Opm
                            const ParseContext& parseContext_,
                            ErrorGuard& errors_,
                            SimulatorUpdate * sim_update_,
-                           const std::unordered_map<std::string, double> * target_wellpi_)
+                           const std::unordered_map<std::string, double> * target_wellpi_,
+                           std::unordered_map<std::string, double>* wellpi_global_factor_ = nullptr)
```
Comment (GitPaean) [inline opm/input/eclipse/Schedule/Schedule.hpp:520]: I think the default value for argument `wellpi_global_factor_` can be removed here.  But I will address it with other potential comments.

**Review by bska — COMMENTED** [review]
I think this makes sense as a first step.  I did consider if there might be some other kind of accumulation structure, but I couldn't come up with any that would satisfy every requirement.  Since `WPIMULT` is completely static information we don't need to trigger any events and handle it after the fact based on dynamic information like we do in the `WELPI` case.

As for removing `= nullptr` we also need to consider how all of this applies in the context of `ACTIONX`.  I don't have that information right now.  Finally, I think I might prefer naming the new parameter `wpimult_*` instead of `wellpi_*`.  Although the effect of both keywords is similar&mdash;multiplying some or all of the connection transmissibility factors by a constant factor&mdash;the risk of confusion is greater if we conflate the keywords in the variable naming.

File: src/opm/input/eclipse/share/keywords/000_Eclipse100/W/WPIMULT [hunk 3]
```diff
@@ -15,23 +15,28 @@
     },
     {
       "name": "I",
-      "value_type": "INT"
+      "value_type": "INT",
+      "default" : -1
```
[the remark this answers is not in the document]
Comment (GitPaean) [reply src/opm/input/eclipse/share/keywords/000_Eclipse100/W/WPIMULT:19]: The default values are negative values. And the commercial simulator treats negative values as defaulted. Adding this change was just to make the function `defaultConCompRec` simpler.

I think I do not have to have this change in this PR now, and just need to rewrite the function `defaultConCompRec`  a little bit.

[the remark this answers is not in the document]
Comment (bska) [reply src/opm/input/eclipse/share/keywords/000_Eclipse100/W/WPIMULT:19]: > I think I do not have to have this change in this PR now, and just need to rewrite the function `defaultConCompRec` a little bit.

I've conferred with other sources and indeed the default values here should be negative.  That said, the `defaultConCompRec` helper could be written as
```C++
auto defaultConCompRec = [](const DeckRecord& wpimult)
{
    return std::all_of(wpimult.begin() + 2, wpimult.end(),
        [](const DeckItem& item)
    {
        return item.defaultApplied(0) || (item.get<int>(0) <= 0);
    }
};
```
provided we also `#include <algorithm>`.  The [`DeckItem::defaultApplied()`](https://github.com/OPM/opm-common/blob/7dd56d2bdec7d7d80e7ded260e812225fe15256c/opm/input/eclipse/Deck/DeckItem.hpp#L53-L54) member function is really helpful in these kinds of circumstances.

[the remark this answers is not in the document]
Comment (GitPaean) [reply src/opm/input/eclipse/share/keywords/000_Eclipse100/W/WPIMULT:19]: Thanks for that. The only thing is that we found during handling this case is that `0` is not treated as `defaulted` in the commercial simulator.

**Comment by GitPaean** [comment]
jenkins build this please

**Comment by GitPaean** [comment]
> As for removing = nullptr we also need to consider how all of this applies in the context of ACTIONX. I don't have that information right now.

yes, that is part that I am not sure about. with the current PR, it will trigger a `runtime_error` .

**Comment by blattms** [comment]
tested and reviewed. Merging.

**Pull request merged**